octodns/provider/etc_hosts.py

```diff
@@ -0,0 +1,125 @@
+        self.log = logging.getLogger('EtcHostsProvider[{}]'.format(id))
+        self.log.debug('__init__: id=%s, directory=%s', id, directory)
+        super(EtcHostsProvider, self).__init__(id, *args, **kwargs)
+        self.directory = directory
+
+    def populate(self, zone, target=False, lenient=False):
+        self.log.debug('populate: name=%s, target=%s, lenient=%s', zone.name,
+                       target, lenient)
+
+        # We never act as a source, at least for now, if/when we do we still
+        # need to noop `if target`
+        return False
+
+    def _apply(self, plan):
+        desired = plan.desired
+        changes = plan.changes
+        self.log.debug('_apply: zone=%s, len(changes)=%d', desired.name,
+                       len(changes))
+        cnames = {}
+        values = {}
+        for record in sorted([c.new for c in changes]):
+            # Since we don't have existing we'll only see creates
+            fqdn = record.fqdn[:-1]
+            if record._type in ('ALIAS', 'CNAME'):
+                # Store cnames so we can try and look them up in a minute
+                cnames[fqdn] = record.value[:-1]
+            elif record._type == 'AAAA' and fqdn in values:
+                # We'll prefer A over AAAA, skipping rather than replacing an
+                # existing A
+                pass
+            else:
+                # If we're here it's and A or AAAA and we want to record it's
+                # value (maybe replacing if it's an A and we have a AAAA
+                values[fqdn] = record.values[0]
+
+        if not isdir(self.directory):
+            makedirs(self.directory)
+
+        filename = '{}hosts'.format(join(self.directory, desired.name))
+        self.log.info('_apply: filename=%s', filename)
+        with open(filename, 'w') as fh:
+            fh.write('##################################################\n')
+            fh.write('# octoDNS ')
+            fh.write(self.id)
+            fh.write(' ')
+            fh.write(desired.name)
+            fh.write('\n')
+            fh.write('##################################################\n\n')
+            if values:
+                fh.write('## A & AAAA\n\n')
+                for fqdn, value in sorted(values.items()):
+                    if fqdn[0] == '*':
+                        fh.write('# ')
+                    fh.write(value)
+                    fh.write('\t')
+                    fh.write(fqdn)
+                    fh.write('\n\n')
+
+            if cnames:
+                fh.write('\n')
+                fh.write('## CNAME (mapped)\n\n')
+                for fqdn, value in sorted(cnames.items()):
+                    # Print out a comment of the first level
+                    fh.write('# ')
+                    fh.write(fqdn)
+                    fh.write(' -> ')
+                    fh.write(value)
+                    fh.write('\n')
+                    # No loop protection :-/
```

Comment:
Should just be able to add the `value` to a set to do a quick form of loop detection here.